cmd/juju/model/advancegeneration.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package model
+
+import (
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/api/modelgeneration"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const (
+	advanceGenerationSummary = "Advances units and/or applications to the next generation."
+	advanceGenerationDoc     = `
+Users need to be able to roll changes to applications in a safe guided 
+processes that controls the flow such that not all units of an HA application 
+are hit at once. This also allows some manual canary testing and provides 
+control over the flow of changes out to the model. 
+
+Examples:
+    juju advance-generation redis
+    juju advance-generation redis/0
+    juju advance-generation redis/0 mysql
+
+See also:
+    add-generation
+    cancel-generation
+    switch-generation
+
+Aliases:
+    advance
+`
+)
+
+// NewAdvanceGenerationCommand wraps advanceGenerationCommand with sane model settings.
+func NewAdvanceGenerationCommand() cmd.Command {
+	return modelcmd.Wrap(&advanceGenerationCommand{})
+}
+
+// advanceGenerationCommand is the simplified command for accessing and setting
+// attributes related to adding model generations.
+type advanceGenerationCommand struct {
+	modelcmd.ModelCommandBase
+
+	api      AdvanceGenerationCommandAPI
+	entities []string
+}
+
+// AdvanceGenerationCommandAPI defines an API interface to be used during testing.
+//go:generate mockgen -package model_test -destination ./advancegenerationmock_test.go github.com/juju/juju/cmd/juju/model AdvanceGenerationCommandAPI
+type AdvanceGenerationCommandAPI interface {
+	Close() error
+	AdvanceGeneration([]string) error
+}
+
+// Info implements part of the cmd.Command interface.
+func (c *advanceGenerationCommand) Info() *cmd.Info {
+	info := &cmd.Info{
+		Name:    "advance-generation",
+		Aliases: []string{"advance"},
+		Purpose: advanceGenerationSummary,
+		Doc:     advanceGenerationDoc,
+	}
+	return jujucmd.Info(info)
+}
+
+// SetFlags implements part of the cmd.Command interface.
+func (c *advanceGenerationCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+}
+
+// Init implements part of the cmd.Command interface.
+func (c *advanceGenerationCommand) Init(args []string) error {
+	if len(args) == 0 {
+		return errors.Errorf("unit and/or application names(s) must be specified")
```

Comment:
I assume the effect of advancing an application is that all units are advanced.

Perhaps down the line there should be some warning if an action at the CLI will auto-complete a generation. I see how advancing the application is useful - if an operator is satisfied with an initial canary unit(s) and they just want to move all others forward.

Reply:
Yes, advancing an application advances all the units.